File: Reverse Substrings Between Each Pair of Parentheses.py
class Solution(object):
    def reverseParentheses(self, s):
        """
        :type s: str
        :rtype: str
        """
        x = []
        for i in range(len(s)):
            if s[i] == '(':
                x.append(i)
            elif s[i] == ')':
                y = s[x[-1]: i+1]
                s = s[:x[-1]] + y[::-1] + s[i+1:]
                x.pop()
                
        rev = ""
        for i in range(len(s)):
            if s[i] != ')' and s[i] != '(':
                rev += s[i]
            
        return rev

File: Reverse Substrings Between Each Pair of Parentheses.cpp
class Solution {
public:
    string reverseParentheses(string s) {
        string ret;
        stack<int>stack;
        int n = s.size();
        vector<int>pair(n, -1);
        for (int i = 0; i < n; i++) {
            if (s[i] == '(') {
                stack.push(i);
            } else if (s[i] == ')') {
                int j = stack.top();
                stack.pop();
                pair[i] = j;
                pair[j] = i;
            }
        }
        
        int i = 0;
        int d = 1;
        while (i < n) {
            if (isalpha(s[i])) {
                ret.push_back(s[i]);
            } else {
                i = pair[i];
                d = -d;
            }
            i += d;
        }
        return ret;
    }
};
